syntax = "proto3";
package grpc_defs;

message LockDataRefId {
    uint64 id = 1;
}

message Empty { }

message ReadRequest {
    LockDataRefId txn = 1;
    string key = 2;
}
message KV {
    string key = 1;
    string value = 2;
}

message WriteRequest {
   LockDataRefId txn = 1;
    KV kv = 2;
}

message Value {
    oneof res {
        string val = 1;
        string error = 2;
    }
}

message WriteError {
    oneof res {
        string error = 1;
        string ok = 2;
    }
}

message KVCollection {
    repeated KV val = 1;
}
message ValueRanged {
    oneof res {
        KVCollection val = 1;
        string error = 2;
    }
}

message Json {
    string inner = 1;
}

message JsonWriteRequest {
    LockDataRefId txn = 1;
    string path = 2;
    Json value = 3;
}

service Replicator {
    // Creates a new transaction
    rpc new_with_time(LockDataRefId) returns (Empty);

    rpc serve_read(ReadRequest) returns (Value);
    rpc serve_range_read(ReadRequest) returns (ValueRanged);

    rpc serve_write(WriteRequest) returns (WriteError);
    rpc commit(LockDataRefId) returns (Empty);
    rpc abort(LockDataRefId) returns (Empty);
}

service MainReplicator {
    rpc create_transaction(Empty) returns (LockDataRefId);
    rpc read(ReadRequest) returns (Json);

    // Todo: better return type for write error/success messages
    rpc write(JsonWriteRequest) returns (Json);

    rpc abort(LockDataRefId) returns (Empty);
    rpc commit (LockDataRefId) returns (Empty);
}